balancer/rls/internal/picker_test.go

```diff
@@ -0,0 +1,671 @@
+// +build go1.10
+
+/*
+ *
+ * Copyright 2020 gRPC authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+package rls
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"testing"
+	"time"
+
+	"github.com/google/go-cmp/cmp"
+	"golang.org/x/sync/errgroup"
+	"google.golang.org/grpc/balancer"
+	"google.golang.org/grpc/balancer/rls/internal/cache"
+	"google.golang.org/grpc/balancer/rls/internal/keys"
+	rlspb "google.golang.org/grpc/balancer/rls/internal/proto/grpc_lookup_v1"
+	"google.golang.org/grpc/metadata"
+)
+
+const (
+	defaultTestTimeout = 1 * time.Second
+	maxAge             = 5 * time.Second
+)
+
+func initKeyBuilderMap(t *testing.T) keys.BuilderMap {
+	t.Helper()
+
+	kb1 := &rlspb.GrpcKeyBuilder{
+		Names:   []*rlspb.GrpcKeyBuilder_Name{{Service: "gFoo"}},
+		Headers: []*rlspb.NameMatcher{{Key: "k1", Names: []string{"n1"}}},
+	}
+	kb2 := &rlspb.GrpcKeyBuilder{
+		Names:   []*rlspb.GrpcKeyBuilder_Name{{Service: "gBar", Method: "method1"}},
+		Headers: []*rlspb.NameMatcher{{Key: "k2", Names: []string{"n21", "n22"}}},
+	}
+	kb3 := &rlspb.GrpcKeyBuilder{
+		Names:   []*rlspb.GrpcKeyBuilder_Name{{Service: "gFoobar"}},
+		Headers: []*rlspb.NameMatcher{{Key: "k3", Names: []string{"n3"}}},
+	}
+	kbm, err := keys.MakeBuilderMap(&rlspb.RouteLookupConfig{
+		GrpcKeybuilders: []*rlspb.GrpcKeyBuilder{kb1, kb2, kb3},
+	})
+	if err != nil {
+		t.Fatalf("Failed to create keyBuilderMap: %v", err)
+	}
+	return kbm
+}
+
+type fakeChildPicker struct {
+	pickCh     chan struct{}
+	pickResult balancer.PickResult
+	err        error
+}
+
+func (p *fakeChildPicker) Pick(_ balancer.PickInfo) (balancer.PickResult, error) {
+	p.pickCh <- struct{}{}
+	return p.pickResult, p.err
+}
+
+// TestPickKeyBuilder verifies the different possible scenarios for
+// forming an RLS key for an incoming RPC.
+func TestPickKeyBuilder(t *testing.T) {
+	kbm := initKeyBuilderMap(t)
+
+	tests := []struct {
+		desc       string
+		rpcPath    string
+		md         metadata.MD
+		wantKey    cache.Key
+		wantResult balancer.PickResult
```

Comment:
wantResult is always `PickResult{}`; remove?  Or will it be expanded later?

(Same in the other test below.)

It would be nice to verify something else is returned, however.  Even if the right picker is invoked, that doesn't mean the RLS picker is properly forwarding its result.

Reply:
Modified the `fakeChildPicker` to be configured with `id` have it return a `fakeSubConn` with the configured id. This makes it possible to verify that the Pick is delegated to the appropriate picker and also that the result of the delegated picker is correctly forwarded by the RLS picker.